src/contrib/ethosu/cascader/proposal_generator.cc

```diff
@@ -0,0 +1,219 @@
+    const CascaderGraph& graph) {
+  std::unordered_map<Part, std::vector<Plan>, ObjectPtrHash, ObjectPtrEqual> plans_by_part;
+  for (const auto& it : plans_by_group) {
+    auto part_group = it.first;
+    auto plans = it.second;
+    int highest_index = 0;
+    Part& index_part = part_group.front();
+    // Determine the Part in the Part group with the highest ID - this will be used to index
+    // the Plans
+    for (const auto& part : part_group) {
+      int pid = graph->GetPartID(part);
+      if (pid >= highest_index) {
+        index_part = part;
+        highest_index = pid;
+      }
+    }
+    plans_by_part[index_part].insert(plans_by_part[index_part].begin(), plans.begin(), plans.end());
+  }
+  return plans_by_part;
+}
+
+Proposal AddPlanToProposal(const Proposal& proposal, const Plan& plan,
+                           const std::unordered_set<TensorConfig>& plan_boundary_configs) {
+  std::vector<Plan> new_plans = proposal->GetPlans();
+  new_plans.push_back(plan);
+  TensorConfigMap new_configs = proposal->GetInputTensorConfigs();
+  // Add input configs from the Plan if they're homed in the cascade region
+  for (const auto& config : plan_boundary_configs) {
+    if (config->GetHomeRegion() == proposal->GetCascadeRegion()) {
+      new_configs[config->GetTensor()] = config;
+    }
+  }
+  // Remove the Plan's output config from the new_configs if it's present because
+  // it won't be an input to the Proposal any more
+  if (new_configs.find(plan->GetOutputConfig()->GetTensor()) != new_configs.end()) {
+    new_configs.erase(plan->GetOutputConfig()->GetTensor());
+  }
+  // The updated memory usage is the memory required to run the Plan plus the
+  // non-local memory that's required in the Proposal at that point in time
+  int new_memory_usage = plan->GetMemoryUsage();
+  for (const auto& it : new_configs) {
+    if (plan_boundary_configs.find(it.second) == plan_boundary_configs.end()) {
+      new_memory_usage += it.first->GetSize();
+    }
+  }
+  new_memory_usage = std::max(new_memory_usage, proposal->GetMemoryUsage());
+  int new_cycles = proposal->GetCycles() + plan->GetCycles();
+  std::vector<Part> new_part_group = proposal->GetPartGroup();
+  new_part_group.insert(new_part_group.end(), plan->GetPartGroup().begin(),
+                        plan->GetPartGroup().end());
+  std::sort(new_part_group.begin(), new_part_group.end());
+  return Proposal(proposal->GetGraph(), new_part_group, new_plans, new_configs,
+                  proposal->GetCascadeRegion(), new_memory_usage, new_cycles);
+}
+
+std::vector<Proposal> GeneratePartialProposals(
+    const CascaderGraph& graph, const HomeMap& home_map, const CascaderOptions options,
+    const std::unordered_map<Part, std::vector<Plan>, ObjectPtrHash, ObjectPtrEqual>& plans_by_part,
+    const std::vector<Part>& partial_proposal_group,
+    std::unordered_map<std::vector<Part>, std::vector<Proposal>>* proposals_by_group) {
```

Comment:
Just curious, what is the reasoning to use a pointer as opposed to references elsewhere ?